### PR TITLE
api parsing

### PR DESCRIPTION
**Pull request by vincentarelbundock**
Here's one way in which ``passport``'s api parsing functionality could be integrated in the existing ``countrycode`` function. The idea is:

1. use separate `parse_google()` and `parse_dstk()` to convert strings
2. 'google' and 'dstk' become valid destination codes
3. for these two destinations, a sanity check forces user to supply 'country.name' as the origin code.

A minimalist implementation can be found in the `api` branch. Usage:

```
> library(countrycode)
> countrycode('United States', 'country.name', 'dstk')
United States
        "USA"
> countrycode(c('United Starts of American', 'Canade'), 'country.name', 'google')
United Starts of American                    Canade
          "United States"                  "Canada"
```

This is really ugly code, and there needs to be error checking and handling of factors, etc. Just a proof of concept of using "google" and "dstk" as destination code. I think that might be a satisfactory solution.

Note: prompted by discussion here https://github.com/alistaire47/passport/issues/5

### CODE REVIEW

File: R/countrycode.R [hunk 1]
```diff
@@ -127,8 +128,12 @@ countrycode <- function(sourcevar, origin, destination, warn = TRUE, nomatch = N
...
     }
 
-    if(!destination %in% colnames(dictionary)){
-        stop("Destination code must correpond to a column name in the dictionary data frame.")
+    if ((destination == 'google') & (origin != 'country.name.en.regex')) {
+        stop('origin must equal "country.name" when the destination is "google".')
+    }
+
+    if ((destination == 'dstk') & (origin != 'country.name.en.regex')) {
+        stop('origin must equal "country.name" when the destination is "google".')
```
Comment (cjyetman) [inline R/countrycode.R]: should be `when the destination is "dstk"`

File: R/parse_api.R [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+		stop('The RJSONIO and RCurl packages must be installed to parse
+			  countries using online APIs (Google or DSTK).')
+	}
+	url = utils::URLencode(paste0(stem, country))
+	dat = RJSONIO::fromJSON(url, simplify = TRUE)
+	if (dat$status == 'OK') {
+		out = dat$results[[1]]$address_components[[1]]$long_name
+	} else {
+		out = NA
+		warning(paste('Failed to parse country name using the Data Science Tool Kit: ', country, '\n'))
```
Comment (cjyetman) [inline R/parse_api.R]: this warning needs to be less specific because it could apply to Google or DSTK responses

**Comment by vincentarelbundock** [comment]
@cjyetman don't spend too much time looking at the specifics here. I think this should be part of a more important refactor that splits "matching" and "parsing" functions, along the lines suggested in the other threat.

I'm just wondering if the call structure makes sense from a user's perspective (i.e., using "google" as destination). Or do we need a separate function for this?

**Comment by cjyetman** [comment]
maybe a `parse = c(NULL, ‘google’,  ‘dark’)` argument?

**Comment by vincentarelbundock** [comment]
But then "destination" isn't clear. Do you want this?

`countrycode("United Starts", "country.name", "iso3c", parse = "google")`

which does a double conversion?

Keeping it to a single translation would make things more predictable with the rest of `countrycode` operation and also reduce the number of arguments.

`countrycode("United Starts", "country.name", "google")`

**Comment by vincentarelbundock** [comment]
Oh oh, or do you mean just internally? NULL would be regex?

**Comment by cjyetman** [comment]
If you're separating the 'parsing' (I'm not really sure what that means) and the 'conversion' process, I was thinking it makes sense that...
1. `parse = NULL` means do not parse--just do things the same as the original default behavior
2. `parse = 'google'` means take the origin vector and 'parse' it, then pass it on to the original default behavior and convert it to destination

but, maybe I'm misunderstanding the purpose of the 'parse' step. As I understand it, with the google API you pass a string--which could be an address, a point of interest, a misspelled country name, etc.--and it makes a best guess as to what country it is/is in, returning a standard country name. Is that what it's doing? so...

```r
countrycode("United States", "country.name", "country.name")
[1] "United States"

countrycode("United States", "country.name", "iso3c")
[1] "USA"

countrycode("United Starts", "country.name", "iso3c")
[1] NA

countrycode("United Starts", "country.name", "country.name", parse = "google")
[1] "United States"

countrycode("United Starts", "country.name", "iso3c", parse = "google")
[1] "USA"

countrycode("Washington Monument", "country.name", "iso3c", parse = "google")
[1] "USA"

countrycode("1600 Pennsylvania Ave NW", "country.name", "iso3c", parse = "google")
[1] "USA"
```

**Comment by cjyetman** [comment]
I guess you could also think of the regex matching as one version of 'parsing', so you could adapt the `origin_regex` argument to be something like...
```r
origin_parsing = c(FALSE, TRUE, 'google', 'dstk')
```
1. `origin_regex = FALSE` = no parsing, only conversion
2. `origin_regex = TRUE` = parse by regex, then convert
3. `origin_regex = 'google'` = parse by google api, then convert
4. `origin_regex = 'dstk'` = parse by dstk api, then convert

**Comment by vincentarelbundock** [comment]
The only thing I was saying is that *internally*, we might want to break off the regex parsing and the API parsing from the main `countrycode` function. Just for code clarity. I don't advocate changing anything that would show up in user experience.

That was the goal of only adding "google" and "dstk" as destination: nothing changes for the user except increased functionality.

But internally, the function is a bit unwieldy, and the regex/api/matching codes perform quite different functions.

**Comment by vincentarelbundock** [comment]
But I really like your examples, actually. This is very cool:

`countrycode("1600 Pennsylvania Ave NW", "country.name", "iso3c", parse = "google")`

**Comment by cjyetman** [comment]
hmmmm... now that I look at that, I wonder if this would be better...
```r
countrycode("Uniited Staates", "country.name", "iso3c")
[1] NA

countrycode("Uniited Staates", "google_parse", "iso3c")
[1] "USA"

countrycode("1600 Pennsylvania Ave NW", "country.name", "iso3c")
[1] NA

countrycode("1600 Pennsylvania Ave NW", "google_parse", "iso3c")
[1] "USA"
```

**Comment by vincentarelbundock** [comment]
And then

`countrycode("lkajsdf", "google", "google")`

would return the default string produced by google.

**Comment by cjyetman** [comment]
> And then `countrycode("lkajsdf", "google", "google")` would return the default string produced by google.

I like it!

**Comment by vincentarelbundock** [comment]
One important problem is that when you submit an address rather than a country name, the Google API returns a numbered list of components (with country in position 6 or 7). When you submit a country, it's in position 1. Province/state would likely be different. Those are not named, so it's pretty tricky to extract the country...

**Comment by vincentarelbundock** [comment]
My bad: we just need to extract the "formatted_address" field. No problem.

**Comment by cjyetman** [comment]
yup...
```r
stem <- 'http://maps.google.com/maps/api/geocode/json?address='
country <- '1600 Pennsylvania Ave NW'
components <- jsonlite::fromJSON(utils::URLencode(paste0(stem, country)))$results$address_components[[1]]
components$long_name[sapply(components$types, function(x) any('country' %in% x))]
[1] "United States"
```

**Comment by cjyetman** [comment]
Just sharing my thought experiment about how the new API should look...

Imagine that you completely separate the 'parsing' and 'converting' process, taking 'parsing' to mean taking unstandardized input and reconciling it to some standardized set of country names or codes, which makes sense because the 'parsing' process is a more heavy duty process whether it's using an external API or doing pattern matching versus doing exact matching and conversion. Additionally, imagine that you also consider the classic regex matching as a form of 'parsing', so that would also be separated out into the 'parsing' process. You would have something like this...
```r
country %>% parse_country(method = 'regex.en') %>% countrycode('country.name', 'iso3c')
country %>% parse_country(method = 'regex.de') %>% countrycode('country.name', 'iso3c')
country %>% parse_country(method = 'google') %>% countrycode('country.name', 'iso3c')
country %>% parse_country(method = 'dstk') %>% countrycode('country.name', 'iso3c')
```
if that were true, `country.name` would actually **not** be the appropriate origin code for the conversion part of the process, because the standardized outputs of the regex, google, and dstk methods would not necessarily be the same, and in this new separated-processes setup, the conversion process would only match on exact matches (it would not do the regex parsing as it has in the past). That would lead to us adding the standardized output of each parsing method as a new origin code, so you would have something like this...
```r
country %>% parse_country(method = 'regex.en') %>% countrycode('regex.en', 'iso3c')
country %>% parse_country(method = 'regex.de') %>% countrycode('regex.de', 'iso3c')
country %>% parse_country(method = 'google') %>% countrycode('google', 'iso3c')
country %>% parse_country(method = 'dstk') %>% countrycode('dstk', 'iso3c')
```
Seeing it like that makes me think it's a bit silly to separate the two processes when you would just be repeating yourself in code for something that's totally obvious given the parse `method` and conversion `origin` options. Which leads me to think that this is really the way to go...
```r
country %>% countrycode('regex.en', 'iso3c')
country %>% countrycode('regex.de', 'iso3c')
country %>% countrycode('google', 'iso3c')
country %>% countrycode('dstk', 'iso3c')
```
Even though 'regex.en', 'regex.de', 'google',  and 'dstk' (and any others that might get added) are not strictly 'origin' codes in the traditional sense of the `coutnrycode` package, we could establish that there are a few 'special' origin codes which cause `countrycode()` to parse/reconcile the input first before attempting to convert it.

just some food for thought

**Comment by vincentarelbundock** [comment]
I'm not sure I understand what you're saying. In your examples, is the `parse_country` exposed to the user? 

I don't think that's a good idea at all, and never had that in mind. The only separation that was proposed was to split off the functions **internally** for code readability. I made some experiments but hate the results, and I'm not sure it's worth pursuing further anyway.

**Comment by cjyetman** [comment]
I know you meant internally only, just trying to understand logically what’s the best way for the user to signal that they want their input to be ‘parsed’ before doing the conversion (which I thought was still up for debate). There are multiple possibilities...
1. use a special ‘origin’ code, but that could be confusing because it’s not really an ‘origin’
2. use a special ‘destination’ code, but that prevents the user from getting the ultimate result in some other destination code
3. use a new ‘parse’ option that is NULL for default behavior or otherwise defines the method of pre-parsing, but that adds a new option and complicates things
3b. if you include regex as a pre-parsing method, you break existing behavior (because currently, option 1 is essentially in place for regex), but if you don’t, parse=NULL still pre-parses if you have origin=‘country.name’, which is a bit confusing
4. other possibilities


Since setting origin=‘country.name’ already triggers a special regex pattern matching process, which is essentially option 1 above, I suggest (if this even happens at all) to stick with that and make google and/or dstk ‘special’ origin codes.

If that was obvious or already decided, sorry... it took me a bit to arrive at that conclusion, so I was just sharing my thought process.

**Comment by vincentarelbundock** [comment]
Ok ok, I get it now. I just woke up, so this is hard for my tiny brain :)

I agree with you. Using "google" and "dstk" as 'special' origin codes makes sense. Then, if those are re-used as destination as well, we do no further processing. If another code is provided, we treaty the parsed string as the input when `origin=country.name`.

**Pull request closed without merging**